### PR TITLE
[Merged by Bors] - refactor(CategoryTheory): make Functor.IsEquivalence a Prop

### PR DESCRIPTION
**Pull request by joelriou**
`Functor.IsEquivalence` is now the combination of `F.Full`, `F.Faithful` and `F.EssSurj`. That such a functor is an equivalence (i.e. there is a quasi-inverse, etc.) is now contained in the definition `F.asEquivalence`. This refactor is to avoid any diamond in the `IsEquivalence` instances (and also for `IsLeftAdjoint` and `IsRightAdjoint` which are also made `Props`). If data is important, one should use `Equivalence` or `Adjunction` instead. This refactor also avoids some duplication of code between the structures `Equivalence` and the former `IsEquivalence`.

`IsLeftAdjoint` and `IsRightAdjoint` are moved to the `Functor` namespace. The namespaces have also been fixed in the file `CategoryTheory.Adjunction.FullyFaithful`. Proofs have been tidied in `CategoryTheory.Monoidal.NaturalTransformation`.

Definitions which relied on `IsLeft/RightAdjoint` have been modified in order to contain the data of an adjunction (e.g. `Reflective`, `MonadicRightAdjoint`, etc.).

---

The most relevant changes are in `CategoryTheory.Equivalence`. Then, most of the fixes are straightforward. Those that are not are mostly in files related to monoidal categories.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by joelriou** [comment]
!bench

**Comment by joelriou** [comment]
!bench

**Comment by joelriou** [comment]
> ```diff
> - ~Mathlib.CategoryTheory.Monoidal.OfHasFiniteProducts   instructions   113.7%
> ```

Note: this is because I have added a new definition `Functor.toMonoidalFunctorOfHasFiniteProducts` which transforms a functor commuting with finite products to a monoidal functor for the monoidal structure given by finite products. I introduced this in order to get a cleaner definition of `cartesianClosedOfEquiv` at https://github.com/leanprover-community/mathlib4/pull/12462/files#diff-0b936081067ad5cf5b6cba85a5895474e77d1d3fa27b3a502700ed5f230ae9e2R415-R416

**Review by dagurtomas — APPROVED** [review]

**Comment by joelriou** [comment]
Thanks very much for having a look at this PR!

> We probably want something like (needs `import Mathlib.CategoryTheory.Adjunction.Opposites`):
> 
> ```lean
> noncomputable def leftAdjointCongr {F G : D ⥤ C} (i : F ≅ G) [F.IsRightAdjoint] :
>     haveI := isRightAdjoint_of_iso i
>     leftAdjoint F ≅ leftAdjoint G :=
> ```

I do not think we want this exact definition because this isomorphism is not unique (if `F` has automorphisms)!

What we could do is generalize the definitions which are used in `leftAdjointUniq`: if `G ⊣ F` and `G' ⊣ F'` are adjunctions, then there are bijections `(G ⟶ G') ≃ (F' ⟶ F)` and `(G ≅ G') ≃ (F' ≅ F)`.

**Comment by dagurtomas** [comment]
> What we could do is generalize the definitions which are used in `leftAdjointUniq`: if `G ⊣ F` and `G' ⊣ F'` are adjunctions, then there are bijections `(G ⟶ G') ≃ (F' ⟶ F)` and `(G ≅ G') ≃ (F' ≅ F)`.

I'm working on this in a separate PR

**Comment by dagurtomas** [comment]
#12625 is the refactor of `leftAdjointUniq`

File: Mathlib/CategoryTheory/EffectiveEpi/Preserves.lean [hunk 1]
```diff
@@ -209,10 +209,10 @@ instance (F : C ⥤ D) [IsEquivalence F] : F.ReflectsEffectiveEpiFamilies where
   reflects {α B} X π _ := by
     let i : (a : α) → X a ⟶ (inv F).obj (F.obj (X a)) := fun a ↦ (asEquivalence F).unit.app _
     have : EffectiveEpiFamily X (fun a ↦ (i a) ≫ (inv F).map (F.map (π a))) := inferInstance
-    simp only [asEquivalence_functor, asEquivalence_inverse, IsEquivalence.inv_fun_map, comp_obj,
-      id_obj, Iso.hom_inv_id_app_assoc, i] at this
+    simp only [inv_fun_map, Iso.hom_inv_id_app_assoc, i] at this
     have : EffectiveEpiFamily X (fun a ↦ (π a ≫ (asEquivalence F).unit.app B) ≫
-        (asEquivalence F).unitInv.app _) := inferInstance
+        (asEquivalence F).unitInv.app _) := by
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/CategoryTheory/EffectiveEpi/Preserves.lean]: This change was unintended, but indeed replacing by `inferInstance` saves 1 heartbeat.

File: Mathlib/CategoryTheory/Equivalence.lean [hunk 2]
```diff
@@ -466,105 +468,140 @@ theorem pow_neg_one (e : C ≌ C) : e ^ (-1 : ℤ) = e.symm :=
...
+/-- A functor is an equivalence of categories if it is faithful, full and
+essentially surjective. -/
+class Functor.IsEquivalence (F : C ⥤ D) : Prop where
+  faithful : F.Faithful := by infer_instance
+  full : F.Full := by infer_instance
+  essSurj : F.EssSurj := by infer_instance
 #align category_theory.is_equivalence CategoryTheory.Functor.IsEquivalence
-#align category_theory.is_equivalence.unit_iso CategoryTheory.Functor.IsEquivalence.unitIso
-#align category_theory.is_equivalence.counit_iso CategoryTheory.Functor.IsEquivalence.counitIso
-#align category_theory.is_equivalence.functor_unit_iso_comp CategoryTheory.Functor.IsEquivalence.functor_unitIso_comp
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/CategoryTheory/Equivalence.lean:545]: Yes, that would be a loop.

[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/CategoryTheory/Equivalence.lean:545]: Are `Prop`-valued loops still bad?

[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/CategoryTheory/Equivalence.lean:545]: Thanks for the reviews! We may try to benchmark this after this PR is merged, but I would rather not attempt this just now.

[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/CategoryTheory/Equivalence.lean:545]: Ok. I'll take a look later today.

**Comment by mattrobball** [comment]
Thanks!

bors delegate+

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**